common/go-utils/src/bin/diff_inventory.rs

```diff
@@ -13,44 +13,39 @@ fn main() {
         std::process::exit(1);
     });
 
-    let upstream_versions: HashSet<String> = list_upstream_artifacts()
+    let upstream_artifacts: HashSet<Artifact> = list_upstream_artifacts()
         .unwrap_or_else(|e| {
             eprintln!("Failed to fetch upstream go versions: {e}");
             std::process::exit(1)
         })
         .into_iter()
-        .map(|a| a.go_version)
         .collect();
 
-    let local_versions: HashSet<String> = Inventory::read(&inventory_path)
+    let inventory_artifacts: HashSet<Artifact> = Inventory::read(&inventory_path)
         .unwrap_or_else(|e| {
             eprintln!("Error reading inventory at '{inventory_path}': {e}");
             std::process::exit(1);
         })
         .artifacts
-        .iter()
-        .map(|r| r.go_version.to_string())
-        .collect();
-
-    let mut added_versions: Vec<String> = upstream_versions
-        .difference(&local_versions)
-        .map(String::to_string)
-        .collect();
-
-    added_versions.sort();
-
-    if !added_versions.is_empty() {
-        println!("Added {}.", added_versions.join(", "));
-    }
-
-    let mut removed_versions: Vec<String> = local_versions
-        .difference(&upstream_versions)
-        .map(String::to_string)
+        .into_iter()
         .collect();
 
-    removed_versions.sort();
-
-    if !removed_versions.is_empty() {
-        println!("Removed {}.", removed_versions.join(", "));
-    }
+    [
+        ("Added", &upstream_artifacts - &inventory_artifacts),
+        ("Removed", &inventory_artifacts - &upstream_artifacts),
+    ]
+    .iter()
+    .filter(|x| !x.1.is_empty())
```

Comment:
This is fine since the tuples are declared right above so it's easy to see what `x.1` is referencing but expanding the tuple makes it a bit clearer:

```suggestion
    .filter(|(_, artifact_diff)| !artifact_diff.is_empty())
```